### ISSUE TITLE
Organisation du codelab

### ISSUE THREAD

**Issue by jabby** [adrienpessu/kotlin-codelab-bttf#2]
Afin de pouvoir avancer sur le codelab, je pense que nous devons définir un nouveau plan. En effet comme tu me le faisais remarquer on a finalement peu de temps au Touraine Tech pour présenter et surtout accompagner les gens sur l'ensemble des possibilités qu'offre Kotlin. 

Un premier pas serait de définir la liste des concepts / syntaxes que l'on veut présenter et ce que l'on va omettre volontairement pour rendre le codelab accessible. Je pense bosser demain soir sur ce sujet afin de te proposer quelque chose. 

J'aurais également une proposition d'organisation générale du codelab pour permettre aux "élèves" de raccrocher sur les exercices même si le précédent n'est pas un franc succès pour eux.

**Comment by jabby** [comment]
Pour les choses que j'aimerais montrer sur le projet : 

- val vs var
- type inference
- Basic Types
- Class et Data class et comment instancier une classe + utilisation du clone
- control flow (if else + try catch avec possibilité d'affecter le résultat)
- when expression (au moins la partie avec un argument)
- null syntax
- Smart casts (pas besoin de recaster si tu fais un *is*)
- Interface
- function (single expression + member function)
- named parameters
- default parameters
- function literals
- late initialization

Y a-t-il d'autres points que tu veux aborder sur le codelab ?

Sinon en terme d'organisation, je te propose de faire des step (branches git) et d'avancer pas à pas. Comme cela, si un élève galère trop, il est possible pour lui de checkout la correction et d'entamer la suite du codelab. Est ce que ça te convient.

**Comment by adrienpessu** [comment]
Je suis d'accord, si on pouvait rajouter le ```companion``` ça serait pas mal.
Je suis d'accord avec toi pour les branches git, je pense aussi qu'il faut en faire suffisamment pour que tout le monde puisse raccrocher à chaque petite étape du codelab. (J'ai déjà eu une mauvaise expérience de codelab ou j'ai décroché à la moitié à cause d'un soucis sur mon IDE et après impossible de tout raccroché)

Pour se concentrer sur Kotlin, on peut venir avec un projet Spring boot déjà et faire plusieurs exercice où il faut implémenter un algo en utilisant tel ou tel feature de Kotlin, histoire de guider les gens.

Qu'en penses-tu?

**Comment by jabby** [comment]
Ok pour le companion. Un oubli de ma part. 

Je suis ok pour plein de branches. J'avoue que c'est pratique quand tu décroches.  

Pour ta proposition sur Kotlin, oui un projet spring boot et faire des exercices pour avancer pas à pas sur une application concrète est une bonne idée. Comme je te disais, j'aime bien l'application de gestion / visualisation de "quotes" venant de Back to the future. Je te propose donc que l'on complète l'application et que l'on découpe en exercice pour faire les diffférents steps.
As tu regardé un peu le lien que je t'ai envoyé sur un autre codelab kotlin ? Certains points sont très intéressants pour l'organisation pure et dure du codelab.

**Comment by adrienpessu** [comment]
J'ai regardé la présentation du codelab, ça à l'air pas mal. Il y a deux trucs que je trouve vraiment super, c'est d'avoir une appli front pour tester le résultat, plutôt que d'utiliser Postman et la deuxième c'est qu'à chaque étape du codelab, il y a des points sur les concepts de Kotlin mise en avant juste avant l'exercice. ça permet de donner un rythme explication/exercice/explication/exercice/... 

Pour les exercices, on peut s'inspirer de Kata de codingdojo genre : [http://codingdojo.org/kata/](http://codingdojo.org/kata/)

**Comment by adrienpessu** [comment]
J'ai découpé le code de la manière suivante : 
- branche ```step0``` pour commencer le codelab
  l'application spring boot avec une base H2, le chargement des données et un endpoint simple
  on pourra dans cette branche montrer un peu de code Kotlin
- branche ```step1```premier exercice
  La même application avec des tests unitaires, les participants devront produire le code correspondant à ces TU. (la conversion de date en chiffre vers date en lettre)
- branche ```step1result``` solution au premier exercice

ça fait peut être gros que premier exercice, qu'est ce que tu en penses?

**Comment by adrienpessu** [comment]
J'ai fini de réorganiser le code en branches de ```step0``` à ```step3``` avec branche result pour chaque step comme vu ensemble. N'hésites pas à faire des remarques. ;)
